installer/ssh_cluster.go

```diff
@@ -0,0 +1,431 @@
+	if c.knownHosts == nil {
+		if err := c.parseKnownHosts(); err != nil {
+			return err
+		}
+	}
+	err := c.knownHosts.HostKeyCallback(hostname, remote, key)
+	if err == knownhosts.HostNotFoundError {
+		fingerprint, _ := c.fingerprintSSHKey(key)
+		hostnameFormatted := hostname
+		remoteAddr := remote.String()
+		if remoteAddr != hostname {
+			hostnameFormatted = fmt.Sprintf("%s (%s)", hostname, remoteAddr)
+		}
+		if c.base.YesNoPrompt(fmt.Sprintf("The authenticity of host '%s' can't be established.\nECDSA key fingerprint is %s.\nAre you sure you want to continue connecting?", hostnameFormatted, fingerprint)) {
+			c.base.SendLog(fmt.Sprintf("Trusting '%s' with key fingerprint %s", hostnameFormatted, fingerprint))
+			file, err := os.OpenFile(c.knownHostsPath(), os.O_WRONLY|os.O_APPEND, os.FileMode(0644))
+			if err != nil {
+				return err
+			}
+			defer file.Close()
+			c.knownHosts.AppendHost(hostname, key, file)
+			return nil
+		}
+	}
+	return err
+}
+
+func (c *SSHCluster) sshConfigForAuth(auth []ssh.AuthMethod) *ssh.ClientConfig {
+	return &ssh.ClientConfig{
+		HostKeyCallback: c.hostKeyCallback,
+		User:            c.base.SSHUsername,
+		Auth:            auth,
+	}
+}
+
+func (c *SSHCluster) testAuthentication(sshConfig *ssh.ClientConfig) bool {
+	for _, ip := range c.TargetIPs {
+		conn, err := ssh.Dial("tcp", ip+":22", sshConfig)
+		if err != nil {
+			return false
+		}
+		conn.Close()
+	}
+	return true
+}
+
+func (c *SSHCluster) findSSHAuth() error {
+	c.base.SendLog("Detecting authentication")
+
+	testAndAddAuthMethod := func(a ssh.AuthMethod) bool {
+		sshConfig := c.sshConfigForAuth([]ssh.AuthMethod{a})
+		if c.testAuthentication(sshConfig) {
+			c.base.SendLog("Verified authentication")
+			c.sshAuth = sshConfig.Auth
+			c.base.SSHConfig = sshConfig
+			return true
+		}
+		return false
+	}
+
+	testAndAddSigner := func(s ssh.Signer) bool {
+		return testAndAddAuthMethod(ssh.PublicKeys(s))
+	}
+
+	if m := c.sshAgentAuth(); m != nil {
+		if testAndAddAuthMethod(m) {
+			return nil
+		}
+	}
+
+	for _, s := range c.findSSHKeySigners() {
+		if s.Encrypted {
+			if s.publicKey == nil {
+				signer, err := s.Decrypt()
+				if err != nil {
+					continue
+				}
+				if testAndAddSigner(signer) {
+					return nil
+				}
+			} else {
+				if testAndAddSigner(s) {
+					return nil
+				}
+			}
+		} else {
+			signer, err := ssh.NewSignerFromKey(s.key)
+			if err != nil {
+				continue
+			}
+			if testAndAddSigner(signer) {
+				return nil
+			}
+		}
+	}
+
+	if c.base.YesNoPrompt("No working authentication found.\nWould you like to try password auth?") {
+		password := c.base.PromptProtectedInput(fmt.Sprintf("Please enter you password for SSHing into %s", strings.Join(c.TargetIPs, ", ")))
+		if testAndAddAuthMethod(ssh.Password(password)) {
+			return nil
+		}
+	}
+	if c.base.YesNoPrompt("No working authentication found.\nWould you like to provide a private key?") {
```

Comment:
Can we merge these messages together so that you can pick from password or private key auth, currently it's not obvious that key auth will be offered.